### PR TITLE
inplement new database infrastructure with mongoose

### LINKED ISSUE #87: Redesign Backend-Database-structure and communication
As a development team, we want to create a concept for the long-term server structure so that we do not have integration problems with future extensions of the data sets on this basis.

Please consider switching from the node mongoDB driver in the server to mongoose instead. (Hint by @kej-jay)

---

## :white_check_mark: &nbsp; Conditions of satisfaction

- [x] Should include a concept for future data-model for server-communication (includes data abstraction layer)
- [x] Should include a concept (MD -> Wiki) what should be saved for each RPA activity (and how?)
- [x] implement the redesigned database-structure/abstraction layer & tests

### PR DESCRIPTION
**Pull request by kej-jay**
Also created a draft for a mongoose model for the SSoT and fixed minor stuff on the client side to have the test running again.

fixes #87 

## Changes
This issues switches from the mongodb module to a more document oriented approach using mongoose. We will now be able to create models locally and work also based on those attributes.

## Checklist before merge

**Developer's responsibilities**
* [X] **Assign** one or two developers
* [ ] **Change code** if reviewer(s) has/have requested it
* [ ] **Pull request build** has passed
* [X] **tested locally** (in at least chrome & firefox if frontend)
* [X] updated the **documentation**
* [X] added **tests** where necessary

### CODE REVIEW

File: wiki/Database-and-communication.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+## Models
+
+If you are planning on creating a new document type (called model from now on), please create a javascript file in the modules directory on the server side. There you can specify the schema for objects of that type and also create the model and specify, which collection should be used if not that of the name of the model that you specify.
+This can be a bit confusing at first, but I guess you could think of it like this: The schema defines what fields are of what type. The model then assigns this a name and registers it for use. Mongoose will always try to retrieve objects from the remote db from a collection which has the same name as the one specified when creating the model, but this can be overwritten as I mentioned.
+
+For getting into mongoose I can really recommend [the official getting started](https://mongoosejs.com/docs/index.html) as well as a youtube video which can be found [here](https://www.youtube.com/watch?v=5e1NEdfs4is). It creates a first request through express and shows how mongoose can be used in that context. Also the code is linked in the description of that video if you are interested in trying it out for yourself.
```
Comment (SanJSp) [inline wiki/Database-and-communication.md:15]: Really liked this part, thank you! Should be a must-have in every wiki entry regarding a new technology

File: server/models/rpaTaskModel.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+const rpaTaskSchema = new Schema({
+  Application: String,
+  Task: String,
+  Code: String,
+  InputVars: [rpaVariableSchema],
+  Output: rpaVariableSchema
+}, { collection: 'completeCollection' });
+
+mongoose.model('rpa-tasks-variable', rpaVariableSchema);
+mongoose.model('rpa-tasks', rpaTaskSchema);
```
Comment (SanJSp) [inline server/models/rpaTaskModel.js:21]: newline at EOF

File: server/models/singleSourceOfTruthModel.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    successorIds: [mongoose.Types.ObjectId]
+});
+
+const SSoT_Schema = new Schema({
+  robotMetadata: robotMetadataSchema,
+  Task: String,
+  Code: String
+}, { collection: 'completeCollection' });
+
+mongoose.model('SSoTs', SSoT_Schema);
```
Comment (SanJSp) [inline server/models/singleSourceOfTruthModel.js:40]: newline at EOF

File: server/models/singleSourceOfTruthModel.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    type: String,
+    name: String,
+    predecessorIds: [mongoose.Types.ObjectId],
+    successorIds: [mongoose.Types.ObjectId],
+    rpaApplication: String,
+    rpaTask: String,
+    rpaParameters: [inputParameterSchema],
+    outputVariable: Schema.Types.Mixed
+});
+
```
Comment (SanJSp) [inline server/models/singleSourceOfTruthModel.js:20]: I would have expected an element schema here, that is being used as an object in every instruction, marker, ...

What do you think about that?

File: server/models/userAccessObjectModel.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+const {Schema} = mongoose;
+
+const userAccessSchema = new Schema({
+  AccessLevel: String,
+  robotId: mongoose.Types.ObjectId,
+  userId: mongoose.Types.ObjectId
+}, { collection: 'completeCollection' });
+
+mongoose.model('userAccessObjects', userAccessSchema);
```
Comment (SanJSp) [inline server/models/userAccessObjectModel.js:11]: newline at EOF

File: server/models/userAccessObjectModel.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+const mongoose = require('mongoose');
+
+const {Schema} = mongoose;
+
+const userAccessSchema = new Schema({
+  AccessLevel: String,
+  robotId: mongoose.Types.ObjectId,
+  userId: mongoose.Types.ObjectId
+}, { collection: 'completeCollection' });
```
Comment (SanJSp) [inline server/models/userAccessObjectModel.js:9]: Can you explain this to me? In the video tutorial stated in the wiki, the instructor created a user model and a note model. Shouldn't we create a user model and then create an entry in the ssot to link it?

File: server/utils/TestingUtils/MongoDbManager.js [hunk 7]
```diff
@@ -1,56 +1,56 @@
...
+    const Variable = mongoose.model('rpa-tasks-variable');
+  
+    let var1 = new Variable({ Name: 'taste', Type: 'String', Required: true });
+    let var2 = new Variable({ Name: 'isHealthy', Type: 'Boolean', Required: false });
+    let task1 = new Task({ Application: 'My fancy Cookbook', Task: 'Taste Recipe', Code:'Taste Recipe', InputVars: [ var1, var2 ] });
+  
+    let var3 = new Variable({ Name: 'testVar', Type: 'String', Required: true });
+    let var4 = new Variable({ Name: 'anotherVar', Type: 'String', Required: true });
+    let var5 = new Variable({ Name: 'lastInput', Type: 'String', Required: true });
+    let var6 = new Variable({ Name: 'outputVar', Type: 'String', Required: true, InfoText: 'Output variable' });
```
Comment (SanJSp) [inline server/utils/TestingUtils/MongoDbManager.js:49]: Maybe these vars can have better naming. I see, that using them as mock Objects, naming is not that important

**Review by SanJSp — CHANGES_REQUESTED** [review]
Some small changes and a bit of explanation to do. Thank you for your work!

File: wiki/Database-and-communication.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
+# Database and Communication
+
+We are using a MongoDB database to which we connect through our backend using the Mongoose Library/Module.
+
+## How to communicate with the database
+
+If you plan to create a CRED operation which should be available for the frontend, please consider to create a path in the backend, which can then be called from the frontend.
```
[the remark this answers is not in the document]
Comment (kej-jay) [reply wiki/Database-and-communication.md:7]: Yup, i use those interchangeably (Edit/Update)

File: server/models/singleSourceOfTruthModel.js (hunk 4 above)
Comment (kej-jay) [reply server/models/singleSourceOfTruthModel.js:20]: I wasn't too sure about it, but did what I think would fit. This schema should be edited and or built upon in the issue regarding the SSoT saving

File: server/models/userAccessObjectModel.js (hunk 6 above)
Comment (kej-jay) [reply server/models/userAccessObjectModel.js:9]: I actually didn't yet create a user schema, as I think it schould be done in the US to introduce user management.
To be able to test our SSoT saving, I ceated this object/schema, to join a SSoT and a user and giving it the additional attribute of the accesslevel, so we could start using that sooner with the hardcoded value of the user Id in our Code

File: server/controllers/rpaFrameworkCommandsController.js [hunk 9]
```diff
@@ -1,25 +1,29 @@
...
+const mongoose = require('mongoose');
+const rpaModels = require('../models/rpaTaskModel');
 
+// GET /rpa-framework/commands/get-available-applications
 exports.getAvailableApplications = async (req, res) => {
   try {
-    const listOfDistinctApplications = await activityDataRetrieval.getDistinctApplicationsFromDB();
     res.set('Content-Type', 'application/json');
-    res.send(listOfDistinctApplications);
+    mongoose.connect(process.env.MONGODB_URI, { useNewUrlParser: true, useUnifiedTopology: true });
```
Comment (MaximilianV) [inline server/controllers/rpaFrameworkCommandsController.js:8]: If I'm not completely wrong, it's absolutely sufficient to establish the connection to the DB once at startup, i.e., in `server.js`

File: server/controllers/rpaFrameworkCommandsController.js [hunk 10]
```diff
@@ -1,25 +1,29 @@
...
+const rpaModels = require('../models/rpaTaskModel');
 
+// GET /rpa-framework/commands/get-available-applications
 exports.getAvailableApplications = async (req, res) => {
   try {
-    const listOfDistinctApplications = await activityDataRetrieval.getDistinctApplicationsFromDB();
     res.set('Content-Type', 'application/json');
-    res.send(listOfDistinctApplications);
+    mongoose.connect(process.env.MONGODB_URI, { useNewUrlParser: true, useUnifiedTopology: true });    
+    mongoose.model('rpa-tasks').distinct('Application', (err, tasks) => {
```
Comment (MaximilianV) [inline server/controllers/rpaFrameworkCommandsController.js:9]: Same goes for the `.model` call. You are already creating the model object in `models/rpaTaskModel.js` but are not using it.
Usually, the models are created once from the schemas and then imported when needed. (See for example [this tutorial](https://www.robinwieruch.de/mongodb-express-setup-tutorial))

File: server/controllers/rpaFrameworkCommandsController.js (hunk 9 above)
Comment (WolfgangDaniel) [reply server/controllers/rpaFrameworkCommandsController.js:8]: I aggree with Maximilian on this one as my design decision for the server side was inspired by these tutorials ([part1](https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/mongoose) & [part2](https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/routes)) and they do it differently. But perhaps there is a good reason to do it like this. If so what is the reason?

File: wiki/Database-and-communication.md (hunk 1 above)
Comment (WolfgangDaniel) [reply wiki/Database-and-communication.md:15]: I agree with Sandro. Thank you for the great summary! I would perhaps also add this [tutorial](https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/mongoose). It really helped me :)

File: server/models/singleSourceOfTruthModel.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    name: String,
+    predecessorIds: [mongoose.Types.ObjectId],
+    successorIds: [mongoose.Types.ObjectId]
+});
+
+const SSoT_Schema = new Schema({
+  robotMetadata: robotMetadataSchema,
+  Task: String,
+  Code: String
+}, { collection: 'completeCollection' });
```
Comment (WolfgangDaniel) [inline server/models/singleSourceOfTruthModel.js:38]: Probably it is right the way it is but I don't get it. Why does the SSoT_Schema contain Task and Code? Shouldn't it be elements of type array as described in our SSoT_deutsch wiki?

**Review by WolfgangDaniel — CHANGES_REQUESTED** [review]

File: wiki/Database-and-communication.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (SanJSp) [reply wiki/Database-and-communication.md:7]: It's okay for me :)

File: server/models/singleSourceOfTruthModel.js (hunk 11 above)
Comment (SanJSp) [reply server/models/singleSourceOfTruthModel.js:38]: This will be changed with the pr fixing #17

**Review by SanJSp — DISMISSED** [review]

**Review by WolfgangDaniel — APPROVED** [review]

**Review by SanJSp — APPROVED** [review]

**Pull request merged**